src/afrolabs/components/kafka.clj

```diff
@@ -1916,8 +1955,40 @@ Supports a timeout operation."))
       (halt [_] (-comp/halt consumer))
 
       IKTable
-      (ktable-wait-for-catchup [_ topic-partition-offset timeout-period timeout-value]
-        (ktable-atom-wait-for-catchup ktable-state topic-partition-offset timeout-period timeout-value))
+      (ktable-wait-for-catchup [_ topic-partition-offset timeout-duration timeout-value]
+        (ktable-atom-wait-for-catchup ktable-state topic-partition-offset timeout-duration timeout-value))
+      (ktable-wait-until-fully-current [_ timeout-duration timeout-value]
+        (let [timeout-ms (.toMillis ^java.time.Duration timeout-duration)
+              timeout-chan (csp/timeout timeout-ms)
+              result
+              ;; When a ktable is current these `fully-current-notification-ch` will fire often.
+              ;; We will count until we've received two notification channels before returning from this fn.
+              ;; This is done because `fully-current-notification-ch` will likely contain
+              ;; values of having been current before this call, so we need at least one receive
+              ;; to clear it.
+              (loop [current-notifs-received 0
+                     previous-offsets        nil]
+                (let [[v ch] (csp/alts!! [timeout-chan
+                                          fully-current-notification-ch])]
+                  (if (= ch timeout-chan)
+                    timeout-value
+
+                    ;; if we've received a value, we want to receive the _same_ value
+                    ;; from the `fully-current-notification-ch` at least twice
+                    ;; to indicate stability, ie fully caught to a stable point.
+                    (let [same-as-previous?            (= v previous-offsets)
+                          next-current-notifs-received (if same-as-previous?
+                                                         (inc current-notifs-received)
+                                                         0)]
+                      ;; if we've received the same offsets twice, we're good to return
+                      (if (= 2 next-current-notifs-received)
```

Comment:
consumers sometimes bounce on the end of a topic, as new messages are coming in. This is not a high-quality test, but it's ok for what we're doing here.